### PR TITLE
Bug 1948711: jsonnet: apply HA conventions

### PR DESCRIPTION
**Pull request by dgrisonnet**
Apply the following HA conventions to thanos-querier and
prometheus-adapter:
- 2 replicas
- Hard pod anti-affinity on hostname
- Set the maxUnavailable rollout strategy to 25%

* [ ] I added CHANGELOG entry for this change.
* [x] No user facing changes, so no entry in CHANGELOG was needed.

### CODE REVIEW

File: jsonnet/prometheus-adapter.libsonnet [hunk 1]
```diff
@@ -69,6 +70,25 @@ function(params)
       {
         spec+: {
           replicas: 2,
+          affinity+: {
```
Comment (simonpasquier) [inline jsonnet/prometheus-adapter.libsonnet:73]: it's at the wrong level, it should be under `template`.

File: jsonnet/prometheus-adapter.libsonnet [hunk 2]
```diff
@@ -69,6 +70,25 @@ function(params)
...
+            podAntiAffinity+: {
+              // Apply HA conventons
+              preferredDuringSchedulingIgnoredDuringExecution: [
```
Comment (simonpasquier) [inline jsonnet/prometheus-adapter.libsonnet:76]: we need hard affinity so requiredDuring...

File: jsonnet/thanos-querier.libsonnet [hunk 3]
```diff
@@ -266,6 +266,25 @@ function(params)
 
     deployment+: {
       spec+: {
+        affinity+: {
```
Comment (simonpasquier) [inline jsonnet/thanos-querier.libsonnet:269]: same comments

File: jsonnet/thanos-querier.libsonnet [hunk 4]
```diff
@@ -266,8 +266,27 @@ function(params)
...
+          rollingUpdate: {
+            maxUnavailable: '25%',
+          },
+        },
         template+: {
           spec+: {
+            affinity+: {
+              podAntiAffinity+: {
+                // Apply HA conventons
+                requiredDuringSchedulingIgnoredDuringExecution: [
```
Comment (simonpasquier) [inline jsonnet/thanos-querier.libsonnet:280]: we need to remove the `preferredDuringSchedulingIgnoredDuringExecution` key.

File: jsonnet/prometheus-adapter.libsonnet [hunk 5]
```diff
@@ -69,8 +70,27 @@ function(params)
...
+                  requiredDuringSchedulingIgnoredDuringExecution: [
+                    {
+                      labelSelector: {
+                        matchLabels: pa.config.selectorLabels,
+                      },
+                      topologyKey: "kubernetes.io/hostname",
+                    },
+                  ],
+                },
+              },
```
Comment (paulfantom) [inline jsonnet/prometheus-adapter.libsonnet:93]: Should we add the same to kube-prometheus?

File: jsonnet/thanos-querier.libsonnet [hunk 6]
```diff
@@ -266,8 +266,27 @@ function(params)
...
+        strategy+: {
+          // Apply HA conventions
+          rollingUpdate: {
+            maxUnavailable: '25%',
+          },
+        },
         template+: {
           spec+: {
+            affinity+: {
+              podAntiAffinity+: {
```
Comment (paulfantom) [inline jsonnet/thanos-querier.libsonnet:278]:
```suggestion
              podAntiAffinity: {
```

We don't need to extend this, so it can be replaced

Comment (simonpasquier) [reply jsonnet/thanos-querier.libsonnet:278]: should we enhance github.com/thanos-io/kube-thanos to support the configuration of soft/hard affinity like we do in kube-prometheus?

File: jsonnet/prometheus-adapter.libsonnet (hunk 5 above)
Comment (dgrisonnet) [reply jsonnet/prometheus-adapter.libsonnet:93]: I guess we can always make it configurable in kube-promehteus, although we are not running prometheus-adapter in HA mode by default there.

Should we add it to the [anti-affinity addon](https://github.com/prometheus-operator/kube-prometheus/blob/1d2a0e275af97948667777739a18b24464480dc8/jsonnet/kube-prometheus/addons/anti-affinity.libsonnet) or directly to the prometheus-adapter component?

File: jsonnet/thanos-querier.libsonnet (hunk 4 above)
Comment (dgrisonnet) [reply jsonnet/thanos-querier.libsonnet:280]: indeed, I missed the existing anti-affinity configuration

File: jsonnet/prometheus-adapter.libsonnet (hunk 5 above)
Comment (paulfantom) [reply jsonnet/prometheus-adapter.libsonnet]: > we are not running prometheus-adapter in HA mode by default there.

Maybe we should change it then? :)

> Should we add it to the anti-affinity addon or directly to the prometheus-adapter component?

I think adding it to the addon makes the most sense.

File: jsonnet/thanos-querier.libsonnet (hunk 6 above)
Comment (paulfantom) [reply jsonnet/thanos-querier.libsonnet]: Maybe we should create a library for such common functions/snippets? We already have a lib directory in kube-prometheus in https://github.com/prometheus-operator/kube-prometheus/tree/main/jsonnet/kube-prometheus/lib, which we can use for this. We also already have a function for creating anti-affinity rules [here](https://github.com/prometheus-operator/kube-prometheus/blob/1d2a0e275af97948667777739a18b24464480dc8/jsonnet/kube-prometheus/addons/anti-affinity.libsonnet#L17-L28), which we can use as a base for this. Later we can apply this to kube-thanos and CMO as well. WDYT? We would need to be careful not to recreate whole ksonnet though :D

cc @kakkoyun

File: jsonnet/prometheus-adapter.libsonnet (hunk 5 above)
Comment (dgrisonnet) [reply jsonnet/prometheus-adapter.libsonnet]: Sounds good I'll work on an upstream change

Comment (dgrisonnet) [reply jsonnet/prometheus-adapter.libsonnet]: https://github.com/prometheus-operator/kube-prometheus/pull/1095

**Pull request merged**